File: src/cleansia_android/customer-app/src/main/java/cz/cleansia/customer/features/home/HomeTabViewModel.kt
package cz.cleansia.customer.features.home

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import cz.cleansia.core.freshness.Staleness
import cz.cleansia.core.network.ApiError
import cz.cleansia.core.snackbar.SnackbarController
import cz.cleansia.customer.core.catalog.CatalogRepository
import cz.cleansia.customer.core.data.AddressRepository
import cz.cleansia.customer.core.loyalty.LoyaltyRepository
import cz.cleansia.customer.core.memberships.MembershipRepository
import cz.cleansia.customer.core.notifications.NotificationFeedRepository
import cz.cleansia.customer.core.orders.OrderRepository
import cz.cleansia.customer.core.recurring.RecurringBookingRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import javax.inject.Inject
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.joinAll
import kotlinx.coroutines.launch

/**
 * Injection seam for the home screen's six singleton repositories.
 *
 * **Almost no state lives here** — it exists so the screen avoids the EntryPointAccessors
 * pattern. The one exception is [isUserRefreshing], and it is an exception on purpose: see
 * its own note.
 */
@HiltViewModel
class HomeTabViewModel @Inject constructor(
    val addressRepository: AddressRepository,
    val orderRepository: OrderRepository,
    val loyaltyRepository: LoyaltyRepository,
    val membershipRepository: MembershipRepository,
    val catalogRepository: CatalogRepository,
    val recurringBookingRepository: RecurringBookingRepository,
    val notificationFeedRepository: NotificationFeedRepository,
    private val snackbar: SnackbarController,
) : ViewModel() {

    fun refreshCatalog() {
        viewModelScope.launch {
            catalogRepository.refresh().onError { error ->
                if (error !is ApiError.Network) snackbar.showError(error)
            }
        }
    }

    /**
     * Refetch the bell's unread count. Silent on failure — the badge is
     * ambient chrome; a stale (or absent) count self-heals on the next
     * foreground/Home entry and on every inbox open.
     */
    fun refreshNotificationBadge() {
        viewModelScope.launch { notificationFeedRepository.refreshUnreadCount() }
    }

    /**
     * Home entry — a tab switch back, or the process returning to foreground.
     *
     * **The observer is re-attached to an already-STARTED lifecycle on every recomposition, so this runs
     * far more often than once per foreground.** That is why each source is gated on its own freshness
     * watermark: ungated, every tap on Home would cost three network calls.
     * -> /mobile-app/patterns#session-wipe
     */
    fun onResume() {
        refreshIfStale(loyaltyRepository.staleness) { loyaltyRepository.refresh() }
        refreshIfStale(orderRepository.staleness) { orderRepository.refresh() }
        refreshIfStale(membershipRepository.staleness) { membershipRepository.refresh() }
    }

    /**
     * Raised ONLY by the pull gesture.
     *
     * Every other pull-to-refresh screen in these apps binds its spinner to a single
     * repository `loading` flag. Home cannot: it fans out to three independent
     * repositories, and combining their loading flags would mean [onResume]'s ambient
     * pass raises the pull spinner on tab entry — a spinner the user did not ask for,
     * which is the exact invariant the partner dashboard's `isUserRefreshing` exists to
     * protect. Hence one flag, owned by the gesture.
     */
    private val _isUserRefreshing = MutableStateFlow(false)
    val isUserRefreshing: StateFlow<Boolean> = _isUserRefreshing.asStateFlow()

    /**
     * Ungated by staleness, unlike [onResume]: a pull IS the ask, so every source
     * refreshes whether or not its watermark says it is fresh. The iOS twin says the same
     * thing in its own doc comment.
     *
     * One child coroutine per source so a slow one cannot hold up the others, joined so
     * the spinner drops when the last of them lands. Failures stay silent — the cached
     * snapshot is still on screen and is still the best answer available.
     */
    fun pullToRefresh() {
        if (_isUserRefreshing.value) return
        viewModelScope.launch {
            _isUserRefreshing.value = true
            try {
                listOf(
                    launch { loyaltyRepository.refresh() },
                    launch { orderRepository.refresh() },
                    launch { membershipRepository.refresh() },
                ).joinAll()
            } finally {
                _isUserRefreshing.value = false
            }
        }
    }

    /** Each source gets its own coroutine so a slow one cannot hold up the others. */
    private fun refreshIfStale(staleness: Staleness, refresh: suspend () -> Unit) {
        if (!staleness.isStale()) return
        viewModelScope.launch { refresh() }
    }
}

File: src/cleansia_ios/CleansiaCustomer/Sources/Features/Home/HomeTabViewModel.swift
import CleansiaCore
import CleansiaCustomerApi
import Combine
import Foundation

/// Injection-seam VM for `HomeTab` (the `HomeTabViewModel.kt` parity). The home
/// screen observes the customer singleton repositories; this VM mirrors their
/// streams into `@Published` state and hosts the first-paint skeleton gate
/// (`HomeTab.kt:196-215`). The catalog source is the shell's session-lived
/// `BookingViewModel` — the iOS stand-in for Android's shared
/// `CatalogRepository` singleton, so Home and the booking sheet read one cache.
@MainActor
final class HomeTabViewModel: ViewModel {
    @Published private(set) var recentOrders: [CustomerOrderSummary] = []
    @Published private(set) var ordersLoaded = false
    @Published private(set) var ordersLoading = false
    @Published private(set) var recurringTemplates: [RecurringTemplate] = []
    @Published private(set) var recurringLoaded = false
    @Published private(set) var loyaltyAccount: LoyaltyAccount?
    @Published private(set) var loyaltyLoaded = false
    @Published private(set) var membership: MyMembership?
    @Published private(set) var addresses: [SavedAddress] = []
    @Published private(set) var selectedAddressId: String?
    @Published private(set) var packages: [CatalogPackage] = []
    @Published private(set) var firstPaintReady = false

    private let orderRepository: OrderRepository
    private let recurringRepository: RecurringBookingRepository
    private let loyaltyRepository: LoyaltyRepository
    private let membershipRepository: MembershipRepository
    private let savedAddressRepository: SavedAddressRepository
    private let catalogSource: BookingViewModel
    private let snackbar: SnackbarController
    private var cancellables: Set<AnyCancellable> = []

    init(
        orderRepository: OrderRepository,
        recurringRepository: RecurringBookingRepository,
        loyaltyRepository: LoyaltyRepository,
        membershipRepository: MembershipRepository,
        savedAddressRepository: SavedAddressRepository,
        catalogSource: BookingViewModel,
        snackbar: SnackbarController
    ) {
        self.orderRepository = orderRepository
        self.recurringRepository = recurringRepository
        self.loyaltyRepository = loyaltyRepository
        self.membershipRepository = membershipRepository
        self.savedAddressRepository = savedAddressRepository
        self.catalogSource = catalogSource
        self.snackbar = snackbar
        super.init()
        orderRepository.$orders.assign(to: &$recentOrders)
        orderRepository.$loaded.assign(to: &$ordersLoaded)
        orderRepository.$loading.assign(to: &$ordersLoading)
        recurringRepository.$templates.assign(to: &$recurringTemplates)
        recurringRepository.$loaded.assign(to: &$recurringLoaded)
        loyaltyRepository.$account.assign(to: &$loyaltyAccount)
        loyaltyRepository.$loaded.assign(to: &$loyaltyLoaded)
        membershipRepository.$current.assign(to: &$membership)
        savedAddressRepository.$addresses.assign(to: &$addresses)
        savedAddressRepository.$selectedId.assign(to: &$selectedAddressId)
        catalogSource.$catalogState
            .map { $0.loadedValue?.packages ?? [] }
            .assign(to: &$packages)
        startFirstPaintWatcher()
    }

    var isPlus: Bool {
        membership?.hasMembership == true
    }

    var hasAnyOrders: Bool {
        !recentOrders.isEmpty
    }

    var displayedAddress: SavedAddress? {
        HomeSections.displayedAddress(addresses, selectedId: selectedAddressId)
    }

    var popularPackages: [CatalogPackage] {
        HomeSections.popularPackages(packages)
    }

    var activeRecurring: [RecurringTemplate] {
        HomeSections.activeRecurring(recurringTemplates)
    }

    /// No membership term: a lapsed membership does not stop a schedule, and hiding a
    /// running schedule from the customer paying for it hides the way to stop it.
    var showRecurringSection: Bool {
        !activeRecurring.isEmpty
    }

    var mostRecentCompleted: CustomerOrderSummary? {
        HomeSections.mostRecentCompleted(recentOrders)
    }

    var recentForDisplay: [CustomerOrderSummary] {
        HomeSections.recentForDisplay(recentOrders)
    }

    var showRecent: Bool {
        HomeSections.showRecent(recent: recentForDisplay, ordersLoaded: ordersLoaded, ordersLoading: ordersLoading)
    }

    var milestoneAccount: LoyaltyAccount? {
        HomeSections.showMilestone(loyaltyAccount) ? loyaltyAccount : nil
    }

    var showSetupRecurringSlide: Bool {
        Self.showSetupRecurringSlide(
            isPlus: isPlus,
            hasRecurringSource: true,
            templatesEmpty: recurringTemplates.isEmpty
        )
    }

    /// The `showSetupRecurringSlide` predicate (`HomeTab.kt:167`). The
    /// wired-source clause is vestigial now that the recurring repository is a
    /// required init param, kept so the showSetupRecurringSlide guard tests
    /// stay meaningful.
    static func showSetupRecurringSlide(isPlus: Bool, hasRecurringSource: Bool, templatesEmpty: Bool) -> Bool {
        isPlus && hasRecurringSource && templatesEmpty
    }

    /// Home entry and every return to the foreground.
    ///
    /// Nothing else on this screen ever re-fetches a cache that already landed:
    /// the shell prefetch runs once per shell entry, and the `.task` warmers
    /// below all guard on "still empty". Loyalty had no warmer here at all, so
    /// points earned after the first fetch only surfaced on the next cold start
    /// (or a Rewards pull-to-refresh) — the reported "only after the next
    /// launch of the app".
    ///
    /// The naive fix is worse than the bug. SwiftUI restarts a `.task` every
    /// time it re-presents the tab root, and `scenePhase` flips to `.active` on
    /// every foreground, so this hook is hot: ungated it would bill three
    /// network calls per tap on the Home tab. Each source is therefore gated on
    /// its own freshness watermark, and a repeated return inside the window
    /// costs nothing.
    ///
    /// Concurrent, so one slow source cannot hold up the others. Silent on
    /// failure — the screen keeps rendering the cached snapshot, so a snackbar
    /// for an ambient background refresh would be noise.
    func refreshStaleSources() async {
        async let loyalty: Void = refreshLoyaltyIfStale()
        async let orders: Void = refreshOrdersIfStale()
        async let membership: Void = refreshMembershipIfStale()
        _ = await (loyalty, orders, membership)
    }

    /// The pull, which deliberately does NOT consult staleness.
    ///
    /// `refreshStaleSources` is ambient — it runs on appear and skips anything inside its freshness
    /// window, which is right for something the customer did not ask for. A pull IS the ask: a customer
    /// who drags the screen down and gets nothing because a watermark says the data is young enough has
    /// been told the gesture does not work. So every source refreshes unconditionally.
    ///
    /// Silent on failure, like its ambient sibling — the cached snapshot stays on screen, and the pull
    /// spinner ending is itself the feedback.
    func pullToRefresh() async {
        async let loyalty: Void = forceRefreshLoyalty()
        async let orders: Void = forceRefreshOrders()
        async let membership: Void = forceRefreshMembership()
        _ = await (loyalty, orders, membership)
    }

    // `refresh()` returns an ApiResult, so these wrap it to Void for the concurrent `async let` —
    // the same shape the staleness-gated siblings below already use. The result is discarded on
    // purpose: a pull that fails leaves the cached snapshot on screen, and the spinner ending is the
    // feedback.
    private func forceRefreshLoyalty() async { _ = await loyaltyRepository.refresh() }

    private func forceRefreshOrders() async { _ = await orderRepository.refresh() }

    private func forceRefreshMembership() async { _ = await membershipRepository.refresh() }

    private func refreshLoyaltyIfStale() async {
        guard loyaltyRepository.staleness.isStale else { return }
        await loyaltyRepository.refresh()
    }

    private func refreshOrdersIfStale() async {
        guard orderRepository.staleness.isStale else { return }
        await orderRepository.refresh()
    }

    private func refreshMembershipIfStale() async {
        guard membershipRepository.staleness.isStale else { return }
        await membershipRepository.refresh()
    }

    /// `LaunchedEffect(Unit) { if (packages.isEmpty) refreshCatalog() }`
    /// (`HomeTab.kt:144-146` + `HomeTabViewModel.kt:39-45`) — surfaces the
    /// snackbar on failure via the codebase-wide `showApiError` convention.
    func refreshCatalogIfNeeded() async {
        guard packages.isEmpty else { return }
        await catalogSource.loadCatalog()
        if case let .error(error) = catalogSource.catalogState {
            snackbar.showApiError(error)
        }
    }

    /// The `LaunchedEffect(Unit) { recurringRepo.refresh() }` parity
    /// (`HomeTab.kt:160-162`) — errors stay silent, as on Android. Skips when the shell
    /// prefetch already landed the templates — this pass only backfills a failed/raced
    /// prefetch. Not gated on membership: a lapsed member's schedules keep generating,
    /// and an unfetched list is a section that cannot appear.
    func refreshRecurring() async {
        guard !recurringRepository.loaded else { return }
        await recurringRepository.refresh()
    }

    var sectionVisibility: HomeSections.SectionVisibility {
        HomeSections.SectionVisibility(
            orderAgain: mostRecentCompleted != nil,
            recurring: showRecurringSection,
            packages: !popularPackages.isEmpty,
            recent: showRecent,
            milestone: milestoneAccount != nil
        )
    }

    /// The 1.5s fallback ceiling (`HomeTab.kt:207-210`) — a slow/failing source
    /// stops blocking and the page renders whatever arrived.
    func runFirstPaintCeiling() async {
        try? await Task.sleep(nanoseconds: 1_500_000_000)
        guard !Task.isCancelled, !firstPaintReady else { return }
        firstPaintReady = true
    }

    /// Flip once when every Home source has landed (recurring only gates Plus
    /// members); never revert for this tab session (`HomeTab.kt:196-203`).
    private func startFirstPaintWatcher() {
        Publishers.CombineLatest(
            Publishers.CombineLatest3($ordersLoaded, $membership, $packages),
            Publishers.CombineLatest($loyaltyLoaded, $recurringLoaded)
        )
        .map { core, extras in
            let (ordersLoaded, membership, packages) = core
            let (loyaltyLoaded, recurringLoaded) = extras
            return HomeSections.firstPaintReady(HomeSections.FirstPaintSources(
                ordersLoaded: ordersLoaded,
                membershipReady: membership != nil,
                packagesReady: !packages.isEmpty,
                loyaltyLoaded: loyaltyLoaded,
                isPlus: membership?.hasMembership == true,
                recurringLoaded: recurringLoaded
            ))
        }
        .filter { $0 }
        .prefix(1)
        .sink { [weak self] ready in self?.firstPaintReady = ready }
        .store(in: &cancellables)
    }
}
